### ISSUE TITLE
Shakespeare Text Diffusion Poor Quality After 1000 Epochs - Analysis & Solutions

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/TinyDiffusionModels#8]
# Shakespeare Text Diffusion Quality Issues After 1000 Epochs

## Problem Summary

After training the Shakespeare text diffusion model for 1000 epochs, the generated samples are producing multilingual garbage output instead of coherent English text. The samples contain random characters, emojis, and fragments from multiple languages rather than Shakespeare-style text.

## Sample Output Quality

**Bad samples from `samples/bad_text/`:**
```
częściej gedachtenynı comprendsдыду "/",🕗 "/") deoarece profiled toată oamenПетерVoltaje mocno deixandoдыдуNaopredic "/")дыду wytrzymaVoltaje🕗 oamenVoltaje remportéбири toată oamen olhando deixando آ toatăдыдуチュアNaoСлаимуществадыду descobriVoltaje raged clutched oamen toată comprendsдыду doubting comprends raisedVoltajeбири "/") deixandoVoltaje nicio oamenii Worry tentando deixando toată toată segurando semnMauricio oamen🕗 boîtier dumnesetWeight raged🕗Voltaje endangeringNaoオレ imagines imagined jeopardi imagines comprendsбириVoltajelepiej ricevutoहर comprendsдыду gedachten攞 frightened semnToda doubtingFromNibVoltaje oamen emphas🕗 părбириVoltajeVoltajegleicheVoltaje chargeur einiges oamenMej mocnoСла imagineslepší拡Voltaje frightenedNao păr comprendsдыду weakened🕗オレVoltaje adaptée faudraVoltaje
```

**Recent samples from `samples/sample_0.txt`:**
```
'
 seguintes


CyfeiriadauBO
BO рэ'";



;

Cyfeiriadau



pageXbackground




フォローδας
πό

'
InputBorder



édr slou'


pper




Cyfeiriadau'),
', больших;the
```

### 1. Embedding Space Mismatch (Critical Issue)
**Location:** `src/shakespeare.py:235-236`
```python
tokenizer = AutoTokenizer.from_pretrained(args.model_id)  # Gemma-2b-it
lm_model = AutoModelForCausalLM.from_pretrained(args.model_id).to(device)
embed_matrix = lm_model.get_input_embeddings().weight.detach().to(device)
```

**Problem:** Using Gemma-2b-it embeddings for Shakespeare text creates fundamental incompatibility:
- Gemma wasn't trained on early modern English
- Vocabulary/style mismatch is severe 
- Denoising moves embeddings to regions that don't map to meaningful Shakespeare tokens

### 2. Naive Embedding-to-Token Decoding
**Location:** `src/shakespeare.py:149-152`
```python
emb_norm = F.normalize(embed_matrix, dim=1)  # (V, dim)
x_norm = F.normalize(x, dim=2)               # (B, L, dim)
sims = torch.matmul(x_norm, emb_norm.T)      # (B, L, V)
tokens = sims.argmax(dim=-1)                 # (B, L)
```

**Problem:** Simple cosine similarity produces multilingual garbage:
- No learned mapping from denoised embeddings back to coherent tokens
- High-dimensional embedding space is too sparse for direct similarity matching
- Denoised embeddings land in regions between valid token embeddings

### 3. Architecture Limitations
**Location:** `src/shakespeare.py:46-59`
```python
class TinyTransformer(nn.Module):
    def __init__(self, dim, n_heads=4, depth=3):  # Only 3 layers\!
```

**Problems:**
- TinyTransformer (3 layers, 4 heads) lacks capacity for complex embedding relationships
- Simple linear time embedding insufficient for 1000 timesteps
- No positional encoding for sequence structure

### 4. Training Configuration Issues
**Location:** `deployment/configs/shakespeare-training.yaml:10`
```yaml
args: ["--train", "--epochs", "10000", "--batch_size", "32", "--ckpt", "gs://text-diffusion/diffusion/outputs/model/text-model.pth"]
```

**Problems:**
- Default sequence length 64 too short for coherent text generation
- Batch size 32 may limit gradient stability
- 1000 epochs may be insufficient for embedding space convergence

#### 1. Switch to Discrete Token Diffusion
**Rationale:** Eliminate embedding space issues entirely
- Operate directly on token IDs rather than embedding space
- Use categorical diffusion with learned noise schedules (e.g., D3PM, SUNDAE)
- No embedding-to-token mapping required

#### 2. Use Domain-Matched Model
**Rationale:** Better vocabulary alignment
- Replace Gemma with GPT-2 or model trained on historical text
- Fine-tune embeddings on Shakespeare corpus first
- Ensure vocabulary contains appropriate early modern English tokens

#### 3. Neural Embedding Decoder
**Rationale:** Learn proper embedding-to-token mapping
- Train learnable decoder network from embeddings to logits
- Replace cosine similarity with neural mapping (e.g., MLP + softmax)
- Joint training with diffusion model

#### 4. Architectural Improvements
- Increase TinyTransformer depth to 6-12 layers
- Add sinusoidal positional embeddings
- Better time conditioning (learned embeddings + cross-attention)
- Increase attention heads to 8-16

#### 5. Training Modifications
- Increase sequence length to 256+ tokens
- Train for 10K+ epochs with learning rate scheduling
- Larger batch sizes (64-128) for stability
- Add gradient clipping and EMA

#### 6. Data Preprocessing
- Better tokenization strategy for Shakespeare text
- Sequence overlap during chunking
- Data augmentation techniques

#### 7. Advanced Techniques
- Classifier-free guidance for better quality
- Multiple diffusion objectives (reconstruction + perplexity)
- Progressive training (shorter sequences → longer sequences)

## Recommended Immediate Action

**Priority 1:** Implement discrete token diffusion to eliminate the embedding space mismatch issue entirely. This addresses the root cause of the multilingual garbage output.

**Priority 2:** If continuing with embedding diffusion, switch to a domain-appropriate base model (GPT-2) and implement a learned decoder network.

## Technical Details

- **Training Config:** 10000 epochs, batch_size 32, n1-highmem-8 + T4 GPU
- **Model:** TinyTransformer (3 layers, 4 heads) + Gemma-2b-it embeddings
- **Diffusion:** 1000 timesteps, linear beta schedule
- **Git Commit:** Current main branch
- **Deployment Job:** shakespeare-training via Vertex AI

## Code References

- Main implementation: `src/shakespeare.py`
- Embedding extraction: `src/shakespeare.py:235-236`
- Decoding logic: `src/shakespeare.py:149-152`
- Model architecture: `src/shakespeare.py:46-59`
- Training config: `deployment/configs/shakespeare-training.yaml`
